Optuna Integration for automated Hyper parameter search

Anindyadeep opened the pull request:
### Description
This PR targets to add a new feature of automated hyperparameter search using [Optuna](https://optuna.org/).  Additionally, it also introduces a new spec for doing a hyperparameter search using three ways.

- [x] Hyperparameter search using Optuna
- [ ] Hyperparameter search using GPT API using the TaskType and complex prompting
- [ ] Hybrid search using GPT and Optuna

This PR solves issue #313 

This is how the `train_model()` function changes from the client's side

```python
from prompt2model.model_trainer import GenerationModelTrainer
from pathlib import Path

trainer = GenerationModelTrainer(
    pre_train_model_name,
    has_encoder=True,
    executor_batch_size=8,
    tokenizer_max_length=1024,
    sequence_max_length=1280,
    device="CPU"
)

args_output_root = Path("result/training_output")
args_output_root.mkdir(parents=True, exist_ok=True)

trained_model, trained_tokenizer = trainer.train_model(
    training_datasets=train_datasets,
    validation_datasets=val_datasets,
    hyperparameter_search_mode="optuna"
)
```

Some additional changes like supporting default hyperparameters as an option is also provided. However that is something needs to be discussed upon.

neubig commented:
Wow, thanks for the contribution @Anindyadeep!

First, a few initial comments:
1. You listed three different things that could be done for hyperparameter search. I would definitely suggest that we split those into three separate PRs (for ease of reviewing). So we can just review the optuna hyperparameter search in this PR.
2. Just to clarify, would you like us to start reviewing this now, or is it still WIP?

Anindyadeep commented:
> Wow, thanks for the contribution @Anindyadeep!
> 
> First, a few initial comments:
> 
> 1. You listed three different things that could be done for hyperparameter search. I would definitely suggest that we split those into three separate PRs (for ease of reviewing). So we can just review the optuna hyperparameter search in this PR.
> 2. Just to clarify, would you like us to start reviewing this now, or is it still WIP?
> 3. It seems that this is not passing formatting checks. I would suggest that you run pre-commit checks, as detailed [here](https://github.com/neulab/prompt2model/blob/main/CONTRIBUTING.md#developer-installation).

- Yes, I intend to split that into three different PRs, I just have mentioned it here, will remove that Todo eventually.

- I also mentioned @viswavi to start a quick review and would appreciate it if you also provide a high-level review of this. (However the work is still in progress, I have to test the working of the hyperparameter search since, my device does not have that much GPU) 

- As I have mentioned, currently I have a gap in making changes and testing those, so I just wanted to add a PR to get your views as well as run the branch inside colab. We can make this a draft PR for now. However, please do provide your thoughts on the changes I am making eventially.

neubig commented:
Sounds great, I'll take a look when I have a chance.

viswavi commented on prompt2model/param_selector/generate.py:
```diff
@@ -0,0 +1,39 @@
+# - Dynamic initialization of hyperparamter range from task type and complexity
+# - Using LLM to suggest hyperparameter
+
+class AutomamatedParamSelector:
```
Typo in name

```suggestion
class AutomatedParamSelector:
```

viswavi commented on prompt2model/param_selector/generate.py:
```diff
@@ -0,0 +1,39 @@
+"""This module provides a dummy trainer for testing purposes."""
```
Please update the docstring to be more accurate. Moreover, I don't think the file name `prompt2model/param_selector/generate.py` is descriptive. Maybe instead call it `prompt2model/param_selector/optuna_selector.py`?

If you can be more specific about the algorithm used underneath Optuna (e.g. Bayesian optimization) then we can be even more specific in the ffile name.

viswavi commented:
Hi @Anindyadeep, I made a quick pass through this and generally it looks very good. Thank you for the quick work! I've left 2 minor comments in the PR. After addressing those, can you potentially clean up the code a little bit?

from the repo root directory, run `pre-commit run --all-files`
and also run `pytest` to make sure this change has not broken any other tests.

After doing this, I will make an in-depth review of the PR.

viswavi commented:
Also, it looks like there may be merge conflicts with `neulab:main`

viswavi commented:
Comment I made to Anindyadeep over DM (copying here for visibility):

"""
I think that this pattern makes sense, but the way I was originally thinking of this was a little different; have a ParamSelector class that wraps the model trainer (rather than being embedded in the model trainer)
so you would pass the model trainer into the ParamSelector class, and then the parameter selector will run this trainer on a bunch of different configurations before ultimately returning a single trained model

I feel that this provides a little more modularity, but I'm open to changing my mind if you can convince me that the other pattern is better 🙂
"""

Anindyadeep commented:
Yeah, @viswavi make sense, will look into that and push another PR with that and pre-recommit all done

Anindyadeep commented:
Hey @viswavi, added some more changes from our previous discussion. Currently checks inside precommit is passing. However, for certain reasons and changes, tests are breaking. We might need to do some discussion on this and I can re iterate on the commits.

viswavi commented on prompt2model/model_trainer/generate.py:
```diff
@@ -32,6 +34,7 @@ def __init__(
         executor_batch_size: int = 10,
         tokenizer_max_length: int = 512,
         sequence_max_length: int = 1024,
+        device: Optional[str] = None,
```
We prefer to use union types, which would set this as:
```suggestion
        device: str | None = None,
```

To support this, you will also have to add
`from __future__ import annotations  # noqa FI58`

In the first line of this file.

viswavi commented on prompt2model/model_trainer/generate.py:
```diff
@@ -49,6 +52,9 @@ def __init__(
                 allowed to generate when being evaluated on validation dataset.
                 Note that sequence_max_length might be scaled in the ModelExecutor
                 if it exceeds the model's max_embedding.
+            device: The device in which the operations should happen. It is ideal to
+                use GPU. Hence this is Optional, where if no value is given, it will
+                be assumed that GPU is being used.
```
Suggestion for making this a little clearer:
```suggestion
            device: The device in which the operations should happen. If not set, we
                will infer the right device to use ("cuda:0" if cuda is available, 
                otherwise "cpu").
```

viswavi commented on prompt2model/param_selector/search_with_optuna.py:
```diff
@@ -0,0 +1,188 @@
+from prompt2model.model_trainer import BaseTrainer
+from prompt2model.param_selector.base import ParamSelector
+
+# TODO:
+# - User tweaking hyperparameter
+# - Dynamic initialization of hyperparamter range from task type and complexity
+# - Using LLM to suggest hyperparameter
```
Let's not keep TODOs in the code. Instead, please create issues to track these suggestions 😄

Anindyadeep replied:
Sure, will be putting issues from the next set of commits

viswavi commented on prompt2model/param_selector/search_with_optuna.py:
```diff
@@ -0,0 +1,188 @@
+        Returns:
+            Returns the transformers.Trainer class with the optimal hyperparameters
+        """
+        # TODO:
+        # - Find the industry standards for default values spec
+        # - More asserts for other keys. Example checking the min or max values
```
Delete the TODOs and add as issues. Also, we/others can help choose these default values over time.

viswavi commented on prompt2model/model_trainer/base.py:
```diff
@@ -25,6 +25,28 @@ def __init__(self, pretrained_model_name: str):
         )
         self.wandb = None
 
+    # we can also have another method called register_trainer
+    # or something like this
```
Feel like this is unnecessary unless you have a particular reason for it.

```suggestion
```

Anindyadeep replied:
Will remove this, all of those comments were some of my personal issues / todos but will follow the standards

viswavi commented on prompt2model/model_trainer/base.py:
```diff
@@ -25,6 +25,28 @@ def __init__(self, pretrained_model_name: str):
+    # or something like this
+    @property
+    def trainer(self) -> transformers.Trainer:
+        """Access the hugging face trainer class.
```
```suggestion
        """Access the Hugging Face trainer class.
```

viswavi commented on prompt2model/model_trainer/base.py:
```diff
@@ -25,6 +25,28 @@ def __init__(self, pretrained_model_name: str):
+        """Access the hugging face trainer class.
+
+        Returns:
+            transformers.Trainer: The hugging face trainer class
```
```suggestion
            transformers.Trainer: The Hugging Face trainer class
```

viswavi commented on prompt2model/model_trainer/base.py:
```diff
@@ -25,6 +25,28 @@ def __init__(self, pretrained_model_name: str):
+        def model_init():
+            return self.model
+
+        return transformers.Seq2SeqTrainer(
+            model=self.model,
```
I think this is ok for now, but I'm wondering if we can instead use the actual trainer used by the model. We might need to add a `self.hf_trainer` member variable to support that.

viswavi replied:
If we stick with the current formulation, can we create an issue to record this restriction that we're making in the parameter selector?

Anindyadeep replied:
Well, yes we have an existing problem where we have too many trainers but with different meanings, like `self.trainer` or `trainer` from hugging face and `trainer` from `BaseTrainer` class. So we can have a `self.hf_trainer` can make this more cleaner. Can you bit elaborate how you want to see that in action?

viswavi replied:
Since you refactored the code, it seems that this can all be removed? This property method doesn't seem to be used anywhere.

Anindyadeep replied:
Yes, we can remove this for now.

viswavi commented on prompt2model/model_trainer/base.py:
```diff
@@ -25,6 +25,28 @@ def __init__(self, pretrained_model_name: str):
+        Returns:
+            transformers.Trainer: The hugging face trainer class
+        """
+        # Let's assume we are doing Seq2Seq training here
+        # Also assuming the model has encoder now, just for the sake of not changing the
+        # constructor's defination
```
```suggestion
        # Assume we are doing Seq2Seq training here.
```

viswavi commented on prompt2model/model_trainer/generate.py:
```diff
@@ -236,13 +247,164 @@ def tokenize_dataset(
+    ) -> transformers.Trainer:
+        """Method to do automatic hyperparameter search.
+
+        The hyperparameters in this argument will be the same as the hyperparameter
+        choices in the `train_model()` method. However the hyperparameter_choices
```
I don't think this is entirely accurate, since we're also adding the `optuna` parameters, which are otherwise not given to the `train_model` method. Can you clarify this in the comment?

Anindyadeep replied:
It is coming from the change in the schema that I did. Since we have `static_parameters` and `optuna` as keys, so my thought process that to pass that schema while searching hyperparameters. 

Another reason was that I wanted to search for hyperparameters in the `train_model` method only where the process of optimization will be dependent on how the user provides the schema. If `optuna` is present then we will go for the search else not.

viswavi commented on prompt2model/model_trainer/generate.py:
````diff
@@ -236,13 +247,164 @@ def tokenize_dataset(
+        ```
+        Here all of the keys are optional. Here are the criterions laid out:
+
+        :key `static_hyperparameters`: These are predefined hyperparameters solely
+        for doing model training and not hyperparameter search. if this key is not
````
Are you suggesting that these are hyperparameters that we will use directly for training and not tune via parameter search? If so I think this definition can be improved:
```suggestion
        :key `static_hyperparameters`: These are the subset of hyperparameters that are
        fixed and not tuned via hyperparameter search. if this key is not
```

Anindyadeep replied:
Does this answers this question: https://github.com/neulab/prompt2model/pull/315/files/af6119c63ffb6ce720fa491a415def4c59dc86cf#r1315364617

viswavi commented on prompt2model/model_trainer/generate.py:
```diff
@@ -236,13 +247,164 @@ def tokenize_dataset(
+        for doing model training and not hyperparameter search. if this key is not
+        provided then DEFAULT_HYPERPARAMETERS will be choosen to train model.
+
+        :key `optuna`: The presense of key will signify that the user
```
```suggestion
        :key `optuna`: The presence of this key signifies that the user
```

viswavi commented on prompt2model/model_trainer/generate.py:
```diff
@@ -236,13 +247,164 @@ def tokenize_dataset(
+                "test_size": 0.15,
+            },
+
+            "optuna": {
+                "min_num_train_epochs": 5,
+                "max_num_train_epochs": 10,
+                "save_strategy": ["epoch", "steps", "no"],
+                "evaluation_strategy": ["epoch", "no"],
+                "per_device_train_batch_size": [4, 8, 16, 32],
+                "min_weight_decay": 1e-5,
+                "max_weight_decay": 1e-1,
+                "min_learning_rate": 1e-5,
+                "max_learning_rate": 1e-1,
```
Do we really need to add this in the model trainer? These feel like parameters that could instead be provided directly in the search_with_optuna class.

viswavi commented on prompt2model/param_selector/search_with_optuna.py:
```diff
@@ -0,0 +1,188 @@
+"""This module provides a dummy trainer for testing purposes."""
```
Please update the docstring to be more up-to-date

viswavi commented on prompt2model/param_selector/search_with_optuna.py:
```diff
@@ -0,0 +1,188 @@
+class OptunaParamSelector(ParamSelector):
+    """Uses optuna for searching for hyperparameters"""
+
+    def __init__(self, trainer: BaseTrainer):
```
I think it would be more natural to specify the space of parameters we search against here rather than in `GenerationModelTrainer.train_model`

viswavi replied:
Suggestion: rename this object to `promptmodel_trainer` to avoid clashing with the huggingface trainer member variable name within BaseTrainer. 
```suggestion
    def __init__(self, promptmodel_trainer: BaseTrainer):
```

viswavi commented on prompt2model/param_selector/search_with_optuna.py:
```diff
@@ -0,0 +1,188 @@
+                ),
+            }
+
+        # prepare the training args
+        # we are assuming here that the user will also provide these args with the additional
+        # args for range. Or we can provide an another argument of search_args (dict) that will
+        # tell the user to provide the arguments for doing hyperparamerter search
```
Comments should almost always be complete, grammatical sentences (with proper capitalization and punctuation). I also feel that most of this comment is unnecessarily detailed/confusing.

Anindyadeep replied:
Yeah, currently I am struggling on this part, but will improve on that

viswavi commented on prompt2model/param_selector/search_with_optuna.py:
```diff
@@ -0,0 +1,188 @@
+        trainer.args = training_args
+
+        best_run = trainer.hyperparameter_search(
+            n_trials=5,  # FIXME: Discussion needed, where to put this arg and visibility for the user
```
I think this should be set in the init function for this class.

Anindyadeep replied:
In the init function we only have the `trainer` param, so can I then change the definition?

viswavi commented on prompt2model/utils/config.py:
```diff
@@ -0,0 +1,31 @@
+# because the more the max value, the more it will take time to get the
+# max batch size
+
+MAX_SUPPORTED_BATCH_SIZE = 128
```
This is probably too big. I think something like 12 or 16 makes more sense (since we're assuming that folks will be training BERT-sized models on small GPUs).

Anindyadeep replied:
I am also adding a function that I will put on the next commit, to automatically fix the max batch size. This goes something like this

```python
# this is something we need to know
MAX_SUPPORTED_BATCH_SIZE = 128


def get_max_batch_size(model: nn.Module, input_dim: int, device: str):
    batch_size = MAX_SUPPORTED_BATCH_SIZE

    while True:
        try:
            dummy_tensor = torch.rand(batch_size, input_dim).to(device)
            _ = model(dummy_tensor)
            break
        except RuntimeError as e:
            if "CUDA out of memory" in str(e):
                batch_size //= 2
            else:
                raise e
    return batch_size

```

viswavi replied:
I think we can add this in a follow-up PR? For now we can just use a small batch size by default for simplicity.

Anindyadeep replied:
Yes, we can do it

viswavi commented on prompt2model/param_selector/search_with_optuna.py:
```diff
@@ -0,0 +1,188 @@
+
+    def __init__(self, trainer: BaseTrainer):
+        """Initialize with train/val datasets and a prompt specification"""
+        self.trainer = trainer
```
Suggestion: make this
```suggestion
        self.hf_trainer = promptmodel_trainer.trainer
```

viswavi reviewed (changes requested):
Made a pass. Let's chat more about this implementation, but overall I think this is very nice progress towards having a great parameter selector!

One major thing missing from this, right now, is having unit tests for the new functionality you're proposing. These will take time to right but it will increase the chance that your code will do what you want right away, and it will likely help you find and fix bugs in your implementation.

zhaochenyang20 commented:
@Anindyadeep Thanks so much for your contribution. I am pondering how you can control the training device. I searched against the whole Trainer, but I found that `self.device` is never used? 🤔

I hope that I was wrong, and I guess we can assign the training device, #317 is fixed.

Anindyadeep commented on prompt2model/model_trainer/generate.py:
````diff
@@ -236,13 +247,164 @@ def tokenize_dataset(
+        The hyperparameters in this argument will be the same as the hyperparameter
+        choices in the `train_model()` method. However the hyperparameter_choices
+        parameter should follow this schema as shown in this below example.
+
+        ```python
+        hyperparameter_choice = {
+            "static_hyperparameters": {
+                "output_dir": "./result",
+                "logging_steps": 1,
+                "save_strategy": "no",
+                "num_train_epochs": 10,
+                "per_device_train_batch_size": 100,
+                "warmup_steps": 0,
+                "weight_decay": 0.01,
+                "logging_dir": "./result",
+                "learning_rate": 1e-4,
+                "evaluation_strategy": "epoch",
+                "test_size": 0.15,
+            },
+
+            "optuna": {
+                "min_num_train_epochs": 5,
+                "max_num_train_epochs": 10,
+                "save_strategy": ["epoch", "steps", "no"],
+                "evaluation_strategy": ["epoch", "no"],
+                "per_device_train_batch_size": [4, 8, 16, 32],
+                "min_weight_decay": 1e-5,
+                "max_weight_decay": 1e-1,
+                "min_learning_rate": 1e-5,
+                "max_learning_rate": 1e-1,
+            },
+        }
+        ```
+        Here all of the keys are optional. Here are the criterions laid out:
````
> I don't think this is entirely accurate, since we're also adding the `optuna` parameters, which are otherwise not given to the `train_model` method. Can you clarify this in the comment?

I have mentioned the schema here, User have two choices, they can either go for hyperpartam optimization or they might not. Now here are some of my questions

`static_hyperparameters` are those which are only meant for model training `trainer.train_model()`.

`optuna` when this key is given, the user can have two choices to put values `default` (this means the hp space will be chosen from the default set) or the user can provide the dict where they can provide all the parameters or some selected ones. Now these will be used during the time of optimizing the hyperparams. And the best hyperparameters values will override the existing `static_hyperparameters`

neubig commented:
Hi @viswavi and @Anindyadeep , thanks a lot for working on this! I was wondering if we were still working on this?

Anindyadeep commented:
> Hi @viswavi and @Anindyadeep , thanks a lot for working on this! I was wondering if we were still working on this?

Yes @neubig, I am working on this right now. However I am blocked for some cases, hence paused the work. But I am going to roll out the first iterations soon.

neubig commented:
OK, great! Please tell us if there's anything we can do to help.

Anindyadeep commented:
> OK, great! Please tell us if there's anything we can do to help.

So right now the problem is:

1. Optuna seems to have some of their internal problems with GPU memory management, and hence I was't able to run the code properly.
2. For this, I was requiring the assistance of @viswavi. However, I and he both got busy with some commitments.

What could be the possible solutions
1. Research and see where the bottlenecks are and see if that can be solved. (Status: I am blocked in this)
2. Try to see if there are any alternatives, then we can swap in that with optuna. Here I was trying to explore `ray-tune`, a library by AnyScale. So they seems to have a pretty optimized one. However, I am still doing some trials on that. If things work out, then I was thinking to replace optuna with `ray-tune`.

Anindyadeep commented:
Hi @neubig, sorry was busy earlier for some time. Here is the latest PR I pushed. Added a test for now. May need to discuss with @viswavi to know what more tests need to be added. But I also added the integration for hyperparam selector in CLI. Things are working fine there. Let me know what more needs to be added. 

Thanks

viswavi commented on prompt2model_demo.py:
```diff
@@ -351,46 +353,113 @@ def main():
         while True:
-            line = input("Enter the training batch size:")
+            line = input(
+                "Are you interested to train the model with automatic hyperparamter search? Type 'y' for Yes and 'n' for No. "  # noqa E501
```
typo: 
```suggestion
                "Are you interested to train the model with automatic hyperparameter search? Type 'y' for Yes and 'n' for No. "  # noqa E501
```

viswavi commented on prompt2model/param_selector/search_with_optuna.py:
```diff
@@ -0,0 +1,239 @@
+            )
+
+            result = objective_trainer.train()
+            return result.training_loss
```
We should be selecting hyperparameters based on validation metrics (e.g. validation loss) rather than training metrics

Anindyadeep replied:
As discussed, we can move this to a different PR right?

viswavi reviewed (changes requested):
Left one major comment - we should be doing hyperparameter selection based on validation data rather than training data

viswavi commented on prompt2model_demo.py:
```diff
@@ -351,46 +353,113 @@ def main():
+                    "per_device_train_batch_size": [max_batch_size],
+                },
+            )
+            # FIXME: The batch size selection needs some more smarter way to
+            # avoid cuda out of error and automatic list
+            # creation based on the max size with some intervals
```
Create a Github issue for this instead of keeping a "TODO"-style comment in the code. We want to avoid those.
```suggestion
```

Anindyadeep replied:
So, sorry, I should have checked that before. I will create the GitHub issues for the same.

viswavi commented on prompt2model_demo.py:
```diff
@@ -351,46 +353,113 @@ def main():
+            # FIXME: The batch size selection needs some more smarter way to
+            # avoid cuda out of error and automatic list
+            # creation based on the max size with some intervals
+            train_batch_size = max_batch_size  # need to change this
```
Likewise, we should avoid having these kinds of comments in the code. We can either fix the issue immediately or postpone fixing it for now and create a Github issue to track the issue.

Anindyadeep replied:
Yeah I will create issues for the same.

viswavi commented on prompt2model_demo.py:
```diff
@@ -351,46 +353,113 @@ def main():
-                line_print("Invalid input. Please enter a number.")
-        time.sleep(1)
+        if line == "y":
+            line_print("Moving with training and hyperparameter selection.")
```
Suggestion for cleaning up the language here:

```suggestion
            line_print("Starting training with hyperparameter selection.")
```

viswavi replied:
@Anindyadeep could you accept my suggestion or revise this text otherwise? I think the current language sounds overly casual

viswavi commented on prompt2model_demo.py:
```diff
@@ -351,46 +353,113 @@ def main():
+            train_batch_size = max_batch_size  # need to change this
+
+        else:
+            line_print("Going with normal training")
```
Suggestion for cleaning up the language here:

```suggestion
            line_print("Starting training without hyperparameter selection.")
```

viswavi replied:
@Anindyadeep could you revise this text or accept my suggestion? I think the current language sounds overly casual

viswavi commented on prompt2model/param_selector/search_with_optuna.py:
```diff
@@ -0,0 +1,245 @@
+"""This module provides automatic hyperparameter selection using optuna."""
```
Capitalize Optuna.

```suggestion
"""This module provides automatic hyperparameter selection using Optuna."""
```

viswavi replied:
Please address this comment before merging!

Anindyadeep replied:
Done

viswavi commented on prompt2model/param_selector/search_with_optuna.py:
```diff
@@ -0,0 +1,245 @@
+
+
+class OptunaParamSelector(ParamSelector):
+    """Uses optuna for searching for hyperparameters."""
```
```suggestion
    """Uses Optuna for searching for hyperparameters."""
```

viswavi replied:
Please address this comment before merging!

viswavi commented on prompt2model/param_selector/search_with_optuna.py:
```diff
@@ -0,0 +1,245 @@
+    def __init__(self, trainer: GenerationModelTrainer, n_trials: int):
+        """Initialize with train/val datasets and a prompt specification.
+
+        Args:
+            trainer (BaseTrainer): trainer object from GenerationModelTrainer
+            n_trials (int): The number of trials for conducting hyperparameter search
```
Type info is unnecessary here since we already have type hints in the function definition:

```suggestion
        Args:
            trainer: trainer object from GenerationModelTrainer
            n_trials: The number of trials for conducting hyperparameter search
```

viswavi commented on prompt2model/param_selector/search_with_optuna.py:
```diff
@@ -0,0 +1,245 @@
+    """Uses optuna for searching for hyperparameters."""
+
+    def __init__(self, trainer: GenerationModelTrainer, n_trials: int):
+        """Initialize with train/val datasets and a prompt specification.
```
I think this docstring is incorrect? There are no datasets or prompt spec provided - please update

Anindyadeep replied:
Let me know if this wordings sounds right.

viswavi replied:
Looks good

viswavi commented on prompt2model/param_selector/search_with_optuna.py:
```diff
@@ -0,0 +1,245 @@
+        """Select a model among a set of hyperparameters (given or inferred).
+
+        Args:
+            training_datasets (list[Dataset]): One or more training datasets or
+                subset of the actual dataset for running the trainer.
+            validation_sets (Dataset): A dataset for computing validation metrics.
+            hyperparameter_space (Optional[dict[str, Any]], optional): The set
```
Type info is unnecessary here since we already have type hints in the function definition:

```suggestion
            training_datasets: One or more training datasets or
                subset of the actual dataset for running the trainer.
            validation_sets: A dataset for computing validation metrics.
            hyperparameter_space: The set
```

viswavi commented on prompt2model/utils/config.py:
```diff
@@ -0,0 +1,16 @@
+"""Place to store all the default configurations."""
+MAX_SUPPORTED_BATCH_SIZE = 128
+
+# All the default hyperparameters has to be set to some standardised settings
```
The meaning of this comment is a little unclear to me.

Also the grammar is incorrect, should be "All the default hyperparameters **have** to be set to some standardised settings." (with a period at the end)

Anindyadeep replied:
For now, I am deleting this comment. We can address this in the different PR.

viswavi commented on prompt2model/param_selector/search_with_optuna.py:
```diff
@@ -0,0 +1,245 @@
+                optimal hyperparameter search. Defaults to None.
+
+        Returns:
+            Returns a dict which contains the best hyperparameters
```
Comments should almost always be complete sentences:

```suggestion
            Returns a dict which contains the best hyperparameters.
```

viswavi replied:
Please address this comment before merging!

Anindyadeep replied:
Done

viswavi reviewed (changes requested):
Left one code suggestion and several cosmetic/documentation suggestions. Will make a more detailed pass of the actual algorithmic contents afterwards.

viswavi commented on prompt2model_demo.py:
```diff
@@ -351,46 +353,110 @@ def main():
+            num_trials = input(
+                "Enter the number of trials to conduct hypeparamter search. Press enter to use default value: "  # noqa E501
+            )
+            max_batch_size = input(
+                "Enter the max batch size Press enter to use default: "
+            )
```
Suggestion to add punctuation + specify what the current default is.

```suggestion
            default_batch_size = DEFAULT_HYPERPARAMETERS_SPACE["per_device_train_batch_size"][0]  # noqa E501
            max_batch_size = input(
                "Enter the max batch size. " + \
                f"Press enter to use default: {default_batch_size}"
            )
```

viswavi replied:
If you accept my suggestion above to change the default hyperparameter space setting of `per_device_train_batch_size` from `[2, 4]` (a list) to `2` (a constant), then change line 377 above to:
```
            default_batch_size = DEFAULT_HYPERPARAMETERS_SPACE["per_device_train_batch_size"]  # noqa E501
```

Anindyadeep replied:
Yeah, I overlooked this one. Fixed now

viswavi commented on prompt2model/param_selector/search_with_optuna.py:
```diff
@@ -0,0 +1,245 @@
+            else DEFAULT_HYPERPARAMETERS_SPACE
+        )
+
+    def select_from_spec(
+        self, training_sets: list[Dataset], validation: Dataset, prompt_spec: PromptSpec
+    ) -> tuple[PreTrainedModel, PreTrainedTokenizer]:
+        """Select a model among a set of hyperparameters (given or inferred).
+
+        Args:
+            training_sets: One or more training datasets for the trainer.
+            validation: A dataset for computing validation metrics.
+            prompt_spec: A prompt to infer hyperparameters from.
+
+        Return:
+            A model and tokenizer (with hyperparameters from inferred range).
+        """
+        return super().select_from_spec(training_sets, validation, prompt_spec)
```
Pretty sure this function isn't doing anything, since it's already inherited from the `ParamSelector` superclass

Anindyadeep replied:
Yes, it is not doing anything, should I remove this?

viswavi replied:
Huh, what was the error during pre-commit? I'm wondering if the error is that this function is never used.

If that's the case, then you can leave it here as you've done, and I can purge this method in a future PR.

Anindyadeep replied:
Yeah will discuss that on discord.

Anindyadeep replied:
Done, removed this function and there is no problem.

viswavi commented on prompt2model/utils/config.py:
```diff
@@ -0,0 +1,14 @@
+DEFAULT_HYPERPARAMETERS_SPACE = {
+    "min_num_train_epochs": 10,
+    "max_num_train_epochs": 20,
+    "save_strategy": ["epoch", "steps", "no"],
+    "evaluation_strategy": ["epoch", "no"],
```
Does it make sense to search over these two hyperparameters? Intuitively it seems like these wouldn't affect the model's training (in which case we can keep them as a constant value, e.g. `epoch`)

viswavi replied:
Also, it seems that `evaluation_strategy` is not used in the Optuna parameter tuning, so I think we should definitely be setting that one as a constant. (I also think we should set `save_strategy` as a constant, fyi)

Anindyadeep replied:
Yes, you are right that the current set of strategies would not provide any further optimizations in terms of qualitative search.

However, I feel like I would take another pass to see if some other strategies exist. Because by theory, the save of weights after each epoch would be less effective compared to the saving of checkpoint where the evaluation loss is the least. So, if those kind of strategies exist, then I would try to add those, else will keep this to constant i.e. `epoch`.

viswavi replied:
> because by theory, the save of weights after each epoch would be less effective compared to the saving of checkpoint where the evaluation loss is the least

That makes sense, but do these save strategies actually affect whether we choose the final checkpoint as the one where the evaluation loss is the least?
 
`epoch` just saves the model after a full epoch of training, while `steps` saves a model after a specified number of steps. As I understand it, the latter just saves many more intermediate checkpoints, while the former saves fewer checkpoints.

The final model returned will still be the one at the end of the full training process.

As such, I think it makes sense to use the value `no` here.

Anindyadeep replied:
Yeah that does makes a good point. Making those changes.

viswavi commented on prompt2model/param_selector/search_with_optuna.py:
```diff
@@ -0,0 +1,245 @@
+                "min_weight_decay": hyperparameter_space.get(
+                    "min_weight_decay",
+                    DEFAULT_HYPERPARAMETERS_SPACE.get("min_weight_decay"),
+                ),
+                "max_weight_decay": hyperparameter_space.get(
+                    "max_weight_decay",
+                    DEFAULT_HYPERPARAMETERS_SPACE.get("max_weight_decay"),
+                ),
+                "min_learning_rate": hyperparameter_space.get(
+                    "min_learning_rate",
+                    DEFAULT_HYPERPARAMETERS_SPACE.get("min_learning_rate"),
+                ),
+                "max_learning_rate": hyperparameter_space.get(
+                    "max_learning_rate",
+                    DEFAULT_HYPERPARAMETERS_SPACE.get("max_learning_rate"),
+                ),
+                "min_num_train_epochs": hyperparameter_space.get(
+                    "min_num_train_epochs",
+                    DEFAULT_HYPERPARAMETERS_SPACE.get("min_num_train_epochs"),
+                ),
+                "max_num_train_epochs": hyperparameter_space.get(
+                    "max_num_train_epochs",
+                    DEFAULT_HYPERPARAMETERS_SPACE.get("max_num_train_epochs"),
+                ),
+                "save_strategy": hyperparameter_space.get(
+                    "save_strategy",
+                    DEFAULT_HYPERPARAMETERS_SPACE.get("save_strategy"),
+                ),
+                "evaluation_strategy": hyperparameter_space.get(
+                    "evaluation_strategy",
+                    DEFAULT_HYPERPARAMETERS_SPACE.get("evaluation_strategy"),
+                ),
+                "per_device_train_batch_size": hyperparameter_space.get(
+                    "per_device_train_batch_size",
+                    DEFAULT_HYPERPARAMETERS_SPACE.get("per_device_train_batch_size"),
+                ),
+            }
+            if hyperparameter_space is not None
+            else DEFAULT_HYPERPARAMETERS_SPACE
+        )
```
I wonder if this can't be condensed programatically, since each of these 8 blocks does exactly the same thing. You would just have a constant that specifies the fields to populate.

Anindyadeep replied:
We can condense this with a for loop I guess?

Anindyadeep replied:
Going for that.

viswavi commented on prompt2model/param_selector/search_with_optuna.py:
```diff
@@ -0,0 +1,245 @@
+        tokenizer = self.generation_model_trainer.tokenizer
+
+        # FIXME: Needs discussion where to keep all of these paths
+        best_model_path = "./best_model"
```
We could use `Path("result/trained_model")`, which is where the demo script currently saves the trained model

Anindyadeep replied:
Ohh sure, we can do that. I could have committed the changes, in case if you suggested here.

viswavi commented on prompt2model/param_selector/search_with_optuna.py:
```diff
@@ -0,0 +1,245 @@
+            validation: A dataset for computing validation metrics.
+            hyperparameters: A dictionary of hyperparameter choices.
+
+        If no hyperparameter_space is choosen, then the default hyperparameter_space
```
Suggestion for cleaning up the language here:

```suggestion
        If no hyperparameter_space is specified, then the default hyperparameter_space
```

viswavi commented on prompt2model/param_selector/search_with_optuna.py:
```diff
@@ -0,0 +1,245 @@
+
+        Args:
+            trainer (BaseTrainer): trainer object from GenerationModelTrainer
+            n_trials (int): The number of trials for conducting hyperparameter search
```
Can we elaborate a little bit on what this means? e.g. 
```suggestion
            n_trials (int): The maximum number of parameter configurations to evaluate
                during conducting hyperparameter search.
```

viswavi commented on prompt2model/param_selector/search_with_optuna.py:
```diff
@@ -0,0 +1,245 @@
+        """Select a model among a set of hyperparameters (given or inferred).
+
+        Args:
+            training_datasets (list[Dataset]): One or more training datasets or
+                subset of the actual dataset for running the trainer.
```
Not sure what "subset of the actual dataset" means here. Suggest removing that phrase:

```suggestion
            training_datasets (list[Dataset]): One or more training datasets
                to use for training models.
```

viswavi commented on prompt2model/param_selector/search_with_optuna.py:
```diff
@@ -0,0 +1,245 @@
+                save_strategy=trial.suggest_categorical(
+                    "save_strategy", hyperparameter_space["save_strategy"]
+                ),
+                per_device_train_batch_size=trial.suggest_categorical(
+                    "per_device_train_batch_size",
```
I feel like `per_device_train_batch_size` is not a good hyperparameter to tune, because you should generally just use the maximum batch size that fits in your machine's memory. I suggest removing this (and changing the default in `prompt2model/utils/config.py` to a constant)

Anindyadeep replied:
Yes, we can remove this. However, Suppose I have a batch size of 1, 2, 4. Where 4 is the max batch size. Now qualitative-wise (i.e. in terms of optimization), does greater batch size always provide better optimization? 

To explain with this example:

L1: aggregated validation loss for Batch 1
L2: aggregated validation loss for Batch 2
L3: aggregated validation loss for Batch 4

So is this thing correct always: L1 < L2 < L3. 

Because, we obviously know this, that resource for a batch size of 1 will be lesser than 2 and so on... Since we are doing optimization, I was thinking on that basis. Let me know what are your thoughts here.

viswavi replied:
> So is this thing correct always: L1 < L2 < L3.

I don't think this is correct. Rather, in most cases, L1, L2, and L3 will be approximately identical.

The details are a little technical, but increasing the batch size from 1 to B reduces the variance of the stochastic gradient by 1/B (therefore reducing the number of optimization steps necessary by  a factor of 1/B). However, by increasing the batch size, we're now going to take |D| / B optimization steps rather than |D| steps (where D is the size of the dataset).

In theory, these will cancel out, and increasing the batch size on a fixed dataset will lead to a very similar trained model. In practice, this is desirable, because the time needed to compute a gradient on a batch (of size B) scales sublinearly with the batch size B, until you hit the memory limit of your machine. In practice, they also will likely yield similar performance.

Anindyadeep replied:
That's an awesome explanation. Thanks for the clarification.

viswavi commented on prompt2model/utils/config.py:
```diff
@@ -0,0 +1,14 @@
+    "max_num_train_epochs": 20,
+    "save_strategy": ["epoch", "steps", "no"],
+    "evaluation_strategy": ["epoch", "no"],
+    "per_device_train_batch_size": [2, 4],
```
I feel like `per_device_train_batch_size` is not a good hyperparameter to tune, because you should generally just use the maximum batch size that fits in your machine's memory.
```suggestion
    "per_device_train_batch_size": 2,
```

Anindyadeep replied:
Let's discuss on [this](https://github.com/neulab/prompt2model/pull/315#discussion_r1364499905), and after that will commit the changes.

viswavi commented on prompt2model/param_selector/search_with_optuna.py:
```diff
@@ -0,0 +1,245 @@
+                    low=hyperparameter_space["min_num_train_epochs"],
+                    high=hyperparameter_space["max_num_train_epochs"],
+                ),
+                save_strategy=trial.suggest_categorical(
+                    "save_strategy", hyperparameter_space["save_strategy"]
+                ),
```
As mentioned in my comment in `prompt2model/utils/config.py`, I don't see how this hyperparameter will affect the trained model. If you agree, let's remove it from the set of hyperparameters tuned here, and instead use a fixed default value in `prompt2model/utils/config.py`

Anindyadeep replied:
Yeah will remove this, after doing a bit of additional research on anything better strategy if I missed.

viswavi replied:
As mentioned before, I don't think this is necessary or useful for hyperparameter search.

Anindyadeep replied:
Sure, removed it.

viswavi commented on prompt2model/param_selector/search_with_optuna.py:
```diff
@@ -0,0 +1,245 @@
+            )
+
+            _ = objective_trainer.train()
+            to_optimize = objective_trainer.evaluate()
+            return to_optimize["eval_loss"]
```
Suggestion to make this clearer:
```suggestion
            optimization_targets = objective_trainer.evaluate()
            return optimization_targets["eval_loss"]
```

viswavi commented on prompt2model/param_selector/search_with_optuna.py:
```diff
@@ -0,0 +1,245 @@
+            validation = concatenate_datasets(validation)
+            validation = self.generation_model_trainer.tokenize_dataset(validation)
+
+        def objective(trial: Trial):
```
Can you add a return type hint here?
e.g.
```suggestion
        def objective(trial: Trial) -> float:
```

viswavi reviewed (changes requested):
I finally made a detailed pass through the actual code. Overall I think the shape of this PR is excellent! I made numerous specific comments about the implementation but I think this could be ready to merge after these suggestions/requested changes are addressed!

Anindyadeep commented on tests/param_selector_test.py:
```diff
@@ -0,0 +1,65 @@
+def test_optimize_hyperparameters():
+    """Tests whether the hyperparameter optimization is working correctly or not."""
+    # create a simple training dataset
+    with tempfile.TemporaryDirectory() as _:
```
Yeah sure 😁

viswavi commented on prompt2model/param_selector/search_with_optuna.py:
```diff
@@ -160,7 +154,7 @@ def select_from_hyperparameters(
         tokenizer = self.generation_model_trainer.tokenizer
 
         # FIXME: Needs discussion where to keep all of these paths
```
Remove this comment

Anindyadeep replied:
done

viswavi commented on prompt2model/param_selector/search_with_optuna.py:
```diff
@@ -187,48 +180,13 @@ def select_from_hyperparameters(
+
+        hp_space = {}
+        for key, default_value in DEFAULT_HYPERPARAMETERS_SPACE.items():
+            hp_space[key] = hyperparameter_space.get(key, default_value)
```
What about the situation where the user provides a key in `hyperparameter_space` that is not found in  `DEFAULT_HYPERPARAMETERS_SPACE` (e.g. `max_grad_norm`)?

In the current implementation, this parameter will be ignored. I think there's two ways we can handle this case:
1) Add this hyperparameter to the `hp_space` dictionary anyways (I think this makes the most sense)
2) Log a warning that an unexpected hyperparameter has been provided, then ignore the parameter. This is probably the safest option but will require users to modify the default hyperparameter object if they want to pass in any new hyperparameter.

Anindyadeep replied:
I mean, I am iterating over `DEFAULT_HYPERPARAMETERS_SPACE` and in order to use new key in `hyperparameter_space`, the user has to change the `DEFAULT_HYPERPARAMETERS_SPACE`. So, is't that way the edge case is automatically handled?

viswavi replied:
Yes, but this new key in `hyperparameter_space` will be silently ignored in the current implementation. That may be confusing for a user, who gets no indication that the supplied hyperparameter would be used

Anindyadeep replied:
Ahh got it, so for any new key the user puts, it will just log a warning right?

Anindyadeep replied:
Yes, I can add this.

viswavi reviewed (changes requested):
Requested one other functional change regarding how you're configuring the hyperparameter space for Optuna

Anindyadeep commented:
@viswavi I had to remove the `select_from_base` function from the base otherwise it is failing the tests.

viswavi commented on prompt2model/param_selector/search_with_optuna.py:
```diff
@@ -173,6 +173,15 @@ def _build_hp_space(
+        for key in list(hyperparameter_space.keys()):
+            if key not in default_keys:
+                print(
+                    f"Key {key} is not present in DEFAULT_HYPERPARAMETERS_SPACE. Hence will be ignored",  # noqa
+                    "However, you can expose the key to the Trainer by adding it to DEFAULT_HYPERPARAMETERS_SPACE.",  # noqa
```
Don't use bare `noqa` calls, please use `noqa E501` if you're suppressing a "line too long" warning. However, the first line here doesn't actually seem too long (I see that it's 85 characters which is below the 88 character threshold)

Suggestion:
```
                    f"Key {key} is not present in DEFAULT_HYPERPARAMETERS_SPACE. Hence will be ignored",
                    "However, you can expose the key to the Trainer by adding it to DEFAULT_HYPERPARAMETERS_SPACE.",  # noqa E501
```

viswavi reviewed (approved):
LGTM!

Anindyadeep commented:
Thank you so much @viswavi and professor @neubig for mentoring throughout the project. I learned lot on this process, specifically on perfection and structured approaches. Looking forward to make some more PRs on this amazing project. 

Next one I would like to go for the CLI issue and try to make it better :) 

Thanks once again.
